lib/global-styles.php

```diff
@@ -171,22 +187,50 @@ function gutenberg_experimental_global_styles_get_theme() {
  * @return string CSS rule.
  */
 function gutenberg_experimental_global_styles_resolver( $global_styles ) {
-	$css_rule = '';
-
-	$token    = '--';
-	$prefix   = '--wp' . $token;
-	$css_vars = gutenberg_experimental_global_styles_get_css_vars( $global_styles, $prefix, $token );
-	if ( empty( $css_vars ) ) {
-		return $css_rule;
+	$css_rules = '';
+
+	$selectors = array( 'core' => ':root' );
+
+	// We need to provide a mechanism for blocks to opt-in into this
```

Comment:
@jorgefilipecosta I've tried here what you suggested (pulling the selectors from block.json), so that's working. However, we need a way for blocks to opt-in into this so we know which block.json to read and from where. I'm thinking we'd need some kind of server-side registration for blocks that want to opt-in into this, similar to what we do for dynamic blocks. cc @gziolo as I believe he was doing some work along the same lines.

Reply:
It's probably fine to add those `selectors` field but alone it seems to be confusing, what's the purpose of adding them there?

Reply:
An alternative would be to have a `register_block_selector` function for this as we have for style variations, for example.

It seems to me that this block selector could be a fundamental part of the blocks going forward so I'm inclined to think that it should be somehow connected to the existing block registration process (hence why `block.json` seemed a fine choice) instead of creating a new API. It'd be good if we could avoid _API proliferation_ as well, so I was curious to see what options did we have to read 3rd party block.json in the server.

I'm not sure what other options do we have for this. @youknowriad in light of [this post](https://make.wordpress.org/core/2020/01/23/controlling-the-block-editor/), what would be your thinking about this and the cross-over with existing block & editor settings?

Reply:
Hi @nosolosw, I guess our need for global styles is just a new property as part of block.json. And then we should use whatever mechanism the editor provides for the server to be aware of block.json settings.
I remember one of the objectives of block.json was to make the server aware of client-only blocks. This would allow blocks to be exposed in a rest API for example. @gziolo is it currently possible to access block.json information from the blocks on the server?

Reply:
The purpose of the selector field is to allow styles compilation. If on theme.json we have:
```
{
	"core/paragraph": {
		color: 'red',
	}
}
```
We need to compile that style into something like "p { color: 'red'; }, the selectors allows us to know that "core/paragraph" maps to the selector "p". It is a property of the block so it seems it should be part of block.json and the server needs to be aware of that value so the styles can be compiled.

Reply:
The purpose of the selector field is to allow styles compilation. If on theme.json we have:
```
{
	"core/paragraph": {
		color: 'red',
	}
}
```
We need to compile that style into something like "p { color: 'red'; }, the selectors allows us to know that "core/paragraph" maps to the selector "p". It is a property of the block so it seems it should be part of block.json and the server needs to be aware of that value so the styles can be compiled.

Reply:
I think only blocks that have PHP files are registered on the server at the moment. However we can change it if necessary 😃 There is now new function `register_block_type_from_metadata` that makes it simple. The only question is where to put the loop that would register all blocks and whether we can automate `block.json` detection.

My another comment expands on the topic.

Reply:
> The only question is where to put the loop that would register all blocks and whether we can automate block.json detection.

That's exactly my struggle!

Reply:
I'm sure we can automate, the question should be if it is acceptable from the WP core perspective. I assume based on the previous experience that the preferred approach would be to create a list:

https://github.com/WordPress/wordpress-develop/blob/baa9b2d1f00a4531ba6849ccad8e645c5be15f24/tools/webpack/packages.js#L105-L117

https://github.com/WordPress/wordpress-develop/blob/2b5b39931cb452c0f63df2aa8f7390ff6b89a726/src/wp-settings.php#L271-L281

Reply:
Ah so if a block calls register_block_type_from_metadata, the information part of block.json will become part of the block registry (it is passed to register_block_type). So I guess we don't need to read any block.json files for global styles, we can simply access the block registry and retrieve the selectors from there.

Reply:
I've reverted this to just maintain a static list of selector/block mappings for now. I believe this can land as it is and we should work on that in a separate PR.